Implement linear tree surrogate

bammari opened the pull request:
## Fixes
We generalize the OMLTSurrogate base class to include models other than Keras models. We refactor these changes across the KerasSurrogate class as well. We do this because future interfaces to OMLT models may inherity from the OMLTSurrogate base class without being specific to Keras. These are mostly semantic changes.

## Summary/Motivation:
We add an interface to the OMLT linear-tree capabilities to allow users to implement Linear Model Decision Trees as surrogates within their IDAES models.

## Changes proposed in this PR:
- Generalize OMLTSurrogate 
- Add LinearTreeSurrogate class which inherits from OMLTSurrogate
- Add tests to test this new interface.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

bammari committed:
Implemented Linear Tree Surrogate Class. Refactored messages in OMLT Base Surrogate parent class to be more general. Created tests for linear trees. All tests passing

bammari committed:
Black

bammari committed:
Merge branch 'IDAES:main' into implement-linear-tree-surrogate

bammari committed:
Docstrings cleanup

bammari committed:
Merge branch 'implement-linear-tree-surrogate' of https://github.com/bammari/idaes-pse into implement-linear-tree-surrogate

bpaul4 commented on idaes/core/surrogate/linear_tree_surrogate.py:
```diff
@@ -0,0 +1,232 @@
+
+        This constructor should only be used when first creating the surrogate within IDAES.
+        Once created, this object can be stored to disk using save_to_folder and loaded
+        with load_from_folder
```
```suggestion
        with load_from_folder.
```

bpaul4 commented on idaes/core/surrogate/linear_tree_surrogate.py:
```diff
@@ -0,0 +1,232 @@
+        Once created, this object can be stored to disk using save_to_folder and loaded
+        with load_from_folder
+
+        Args:
+           lt_model: Linear-tree model
+              This is the Linear-tree model that will be loaded.
+           input_labels: list of str
+              The ordered list of labels corresponding to the inputs in the linear-tree model
+           output_labels: list of str
+              The ordered list of labels corresponding to the outputs in the linear-tree model
+           input_bounds: None of dict of tuples
+              Keys correspond to each of the input labels and values are the tuples of
+              bounds (lb, ub)
+           input_scaler: None or OffsetScaler
+              The scaler to be used for the inputs. If None, then no scaler is used
+           output_scaler: None of OffsetScaler
+              The scaler to be used for the outputs. If None, then no scaler is used
```
```suggestion
        Args:
           lt_model: Linear-tree model
              This is the Linear-tree model that will be loaded.
           input_labels: list of str
              The ordered list of labels corresponding to the inputs in the linear-tree model.
           output_labels: list of str
              The ordered list of labels corresponding to the outputs in the linear-tree model.
           input_bounds: None of dict of tuples
              Keys correspond to each of the input labels and values are the tuples of
              bounds (lb, ub).
           input_scaler: None or OffsetScaler
              The scaler to be used for the inputs. If None, then no scaler is used.
           output_scaler: None of OffsetScaler
              The scaler to be used for the outputs. If None, then no scaler is used.
```

bpaul4 reviewed (commented):
These changes look reasonable to me, just a couple of very minor "missing periods" in docstrings but feel free to ignore. I will update the branch and approve the workflow.

bpaul4 committed:
Merge branch 'main' into implement-linear-tree-surrogate

bammari committed:
Update setup.py

Change to most recent version of omlt==1.2.2 to allow for linear-tree capabilities.

bammari commented:
@bpaul4  Thanks for pointing this out. 

It looks like Github CI installs omlt v1.1 in the idaes-pse-dev conda environment when we run actions. Linear tree capabilities were introduced in omlt v1.2.

We have specified omlt==1.1 in setup.py. The most recent version is omlt==1.2.2. I believe changing this should resolve this issue. I have included a new commit that tests this.

bpaul4 commented on setup.py:
```diff
@@ -42,7 +42,7 @@ class ExtraDependencies:
-        "omlt==1.1",  # fix the version for now as package evolves
+        "omlt==1.2.2",  # fix the version for now as package evolves
         "tensorflow",
         "onnx",
```
Do we need to add `lineartree` to this list?

```suggestion
        "onnx",
        "lineartree",
```

bammari committed:
Update setup.py

Add linear-tree as extra dependency.

Co-authored-by: Brandon Paul <86113916+bpaul4@users.noreply.github.com>

bammari committed:
Update setup.py

Change to linear-tree

bammari committed:
Update setup.py

Fix scikit-learn==1.6.1.

ksbeattie committed:
Merge branch 'main' into implement-linear-tree-surrogate

avdudchenko reviewed (commented):
LGMT to me. 

Only note, if we are updating to 1.2.2 for OMLT, can @bammari  test if removing this from onnx (idaes/core/surrogate/onnx_surrogate.py, line 125 ish)  surrogate models? ( I can also make a commit if that's okay).  

```

        # TODO: remove this once new OMLT 1.2 is made available and includes tanh support
        # overrides default available activation functions for ONNX, tanh is not listed in 1.1 but is supported

        omltio.onnx_parser._ACTIVATION_OP_TYPES = [  # pylint: disable=protected-access
            "Relu",
            "Sigmoid",
            "LogSoftmax",
            "Tanh",
        ]
```

ksbeattie commented:
@bammari has graduated, @MAZamarripa will find someone to hand this off to.

ksbeattie commented:
> @bammari has graduated, @MAZamarripa will find someone to hand this off to.

@MAZamarripa any updates on this?

ksbeattie commented:
@MAZamarripa, moving this to the Aug release
